src/server/gc_worker/compaction_filter.rs

```diff
@@ -0,0 +1,341 @@
+        cfg_tracker,
+    });
+}
+
+pub struct WriteCompactionFilterFactory;
+
+impl CompactionFilterFactory for WriteCompactionFilterFactory {
+    fn create_compaction_filter(
+        &self,
+        context: &CompactionFilterContext,
+    ) -> *mut DBCompactionFilter {
+        let gc_context_option = GC_CONTEXT.lock().unwrap();
+        let gc_context = match *gc_context_option {
+            Some(ref ctx) => ctx,
+            None => return std::ptr::null_mut(),
+        };
+        if !gc_context.cfg_tracker.value().enable_compaction_filter {
+            return std::ptr::null_mut();
+        }
+        if gc_context.safe_point.load(Ordering::Relaxed) == 0 {
+            // Safe point has not been initialized yet.
+            return std::ptr::null_mut();
+        }
+
+        let name = CString::new("write_compaction_filter").unwrap();
+        let db = Arc::clone(&gc_context.db);
+        let safe_point = Arc::clone(&gc_context.safe_point);
+
+        let filter = Box::new(WriteCompactionFilter::new(db, safe_point, context));
+        unsafe { new_compaction_filter_raw(name, filter) }
+    }
+}
+
+struct WriteCompactionFilter {
+    bottommost_level: bool,
+    safe_point: Arc<AtomicU64>,
+    db: Arc<DB>,
+
+    write_batch: RocksWriteBatch,
+    key_prefix: Vec<u8>,
+    remove_older: bool,
+
+    // Record the last MVCC delete mark for each level.
+    leveled_tail_deletes: HashMap<usize, Vec<u8>>,
+
+    versions: usize,
+    deleted: usize,
+}
+
+impl WriteCompactionFilter {
+    fn new(db: Arc<DB>, safe_point: Arc<AtomicU64>, context: &CompactionFilterContext) -> Self {
+        // Safe point must have been initialized.
+        assert!(safe_point.load(Ordering::Relaxed) > 0);
+
+        let wb = RocksWriteBatch::with_capacity(Arc::clone(&db), DEFAULT_DELETE_BATCH_SIZE);
+        WriteCompactionFilter {
+            bottommost_level: context.is_bottommost_level(),
+            safe_point,
+            db,
+
+            write_batch: wb,
+            key_prefix: vec![],
+            remove_older: false,
+
+            leveled_tail_deletes: HashMap::default(),
+            versions: 0,
+            deleted: 0,
+        }
+    }
+
+    fn delete_default_key(&mut self, key: &[u8], write_batch: Option<&mut RocksWriteBatch>) {
+        if let Some(write_batch) = write_batch {
+            write_batch.delete(key).unwrap();
+            return;
+        }
+        self.write_batch.delete(key).unwrap();
+        self.flush_pending_writes_if_need();
+    }
+
+    fn delete_write_key(&mut self, key: &[u8], write_batch: Option<&mut RocksWriteBatch>) {
+        if let Some(write_batch) = write_batch {
+            write_batch.delete_cf(CF_WRITE, key).unwrap();
+            return;
+        }
+        self.write_batch.delete_cf(CF_WRITE, key).unwrap();
+        self.flush_pending_writes_if_need();
+    }
+
+    fn flush_pending_writes_if_need(&mut self) {
+        if self.write_batch.data_size() > DEFAULT_DELETE_BATCH_SIZE {
```

Comment:
I think it is better to judge with `count` rather than `data_size`. Because if there are too many keys in one `WriteBatch`, it may cause pipeline write slow